File: box-and-box/resource.mjs
// resource.mjs — Resource Arithmetic, faithful runtime (v0.8)
// The economy the ladder runs on: what an agent can afford, what is conserved, and what a
// repair is worth. A Ledger is a CLOSED double-entry system — value is never created from
// nothing; every spend is a transfer to a sink, every refill a transfer from a treasury — so
// CONSERVATION holds structurally. Depletable resources obey linear logic (used once, no
// duplication, no silent discard); resources marked `reusable` (linear logic's `!` "of-course"
// modality) can be used freely without depletion — exactly the difference between capacity and
// consolidated knowledge in continual learning. The novel bridge PRICES the repair operators:
// deliberation / escalation / reconciliation cost resource, so a known-unknown is resolved only
// when the value of resolving it exceeds the cost (I. J. Good's "Type II" rationality). Laws
// C1–C8. Grounded in Girard 1987 (linear logic); Clearwater/Wellman (market-based control, the
// multi-agent companion); Good 1971 & Russell/Wefald 1989 (metareasoning); and the continual-
// learning stability–plasticity dilemma cast as a capacity-conservation law.

export const SINK = '#sink', TREASURY = '#treasury', FREE = 'free';
const clone = (L) => ({ bal: Object.fromEntries(Object.entries(L.bal).map(([a, r]) => [a, { ...r }])), kind: { ...L.kind } });
export const Ledger = ({ bal = {}, kind = {} } = {}) => ({ bal, kind });
export const balance = (L, acct, res) => (L.bal[acct] && L.bal[acct][res]) || 0;
export const total = (L, res) => Object.values(L.bal).reduce((s, r) => s + (r[res] || 0), 0);
export const INFEASIBLE = Symbol('infeasible');

// the one primitive: move `amt` of `res` between two accounts. Conserves the grand total.
export function transfer(L, res, from, to, amt) {
  if (amt < 0 || balance(L, from, res) < amt) return INFEASIBLE; // no overdraft (the affine floor)
  const M = clone(L); M.bal[from] = M.bal[from] || {}; M.bal[to] = M.bal[to] || {};
  M.bal[from][res] = (M.bal[from][res] || 0) - amt;
  M.bal[to][res] = (M.bal[to][res] || 0) + amt;
  return M;
}
export const spend = (L, acct, res, amt) => transfer(L, res, acct, SINK, amt);       // consume → sink
export const refill = (L, acct, res, amt) => transfer(L, res, TREASURY, acct, amt);   // accrue ← treasury
export const affords = (L, acct, cost) => Object.entries(cost).every(([res, amt]) => balance(L, acct, res) >= amt);
export const feasible = (L, acct, cost) => affords(L, acct, cost);                    // the alethic gate; else the action carries 0̲

// reusable (`!`) vs depletable: `use` depletes a depletable resource, but never a reusable one
export function use(L, acct, res) {
  if (balance(L, acct, res) < 1) return { ok: false, L };
  if (L.kind[res] === 'reusable') return { ok: true, L };          // copy freely — no depletion
  return { ok: true, L: spend(L, acct, res, 1) };                  // linear — consumed exactly once
}

// continual learning: capacity is a CONSERVED resource; knowledge is reusable (`!`).
export const allocate = (L, task, amt) => transfer(L, 'capacity', FREE, 'task:' + task, amt); // free → committed
export function consolidate(L, task, mind = 'mind') {                                          // mint reusable knowledge
  const M = clone(L); M.kind['know:' + task] = 'reusable';
  M.bal[mind] = M.bal[mind] || {}; M.bal[mind]['know:' + task] = 1; return M;
}
export function forget(L, task, mind = 'mind') {                                               // reclaim capacity — only by releasing the knowledge
  const amt = balance(L, 'task:' + task, 'capacity');
  let M = transfer(L, 'capacity', 'task:' + task, FREE, amt);
  if (M === INFEASIBLE) M = clone(L);
  if (M.bal[mind]) M.bal[mind]['know:' + task] = 0;               // the no-free-reclaim tradeoff (stability vs plasticity)
  return M;
}

// PRICING THE REPAIRS (Type II rationality): invoke a repair only if affordable AND worth it.
export const worthwhile = (value, cost) => value >= cost;
export function repair(L, acct, { resource = 'tokens', value, cost }) {
  if (!affords(L, acct, { [resource]: cost })) return { decision: 'cannot-afford', L };
  if (!worthwhile(value, cost)) return { decision: 'skip', L };                                 // act on the current best
  return { decision: 'invoke', L: spend(L, acct, resource, cost) };                             // pay to deliberate / escalate
}

export default { SINK, TREASURY, FREE, Ledger, balance, total, INFEASIBLE, transfer, spend, refill, affords, feasible, use, allocate, consolidate, forget, worthwhile, repair };

File: box-and-box/ports/rust/src/resource.rs
// resource.rs — Resource Arithmetic / linear-logic ledger (faithful port of resource.mjs).
// Laws C1–C8, CB1–CB3. A Ledger is a CLOSED double-entry system — value is conserved.
// Depletable resources obey linear logic; `reusable` resources (`!`) are used freely.

use std::collections::BTreeMap;

pub const SINK: &str = "#sink";
pub const TREASURY: &str = "#treasury";
pub const FREE: &str = "free";

#[derive(Clone, Debug)]
pub struct Ledger {
    // acct -> (res -> amount)
    pub bal: BTreeMap<String, BTreeMap<String, i64>>,
    pub kind: BTreeMap<String, String>,
}

impl Ledger {
    pub fn new() -> Ledger {
        Ledger {
            bal: BTreeMap::new(),
            kind: BTreeMap::new(),
        }
    }
}

pub fn balance(l: &Ledger, acct: &str, res: &str) -> i64 {
    l.bal.get(acct).and_then(|r| r.get(res)).copied().unwrap_or(0)
}

pub fn total(l: &Ledger, res: &str) -> i64 {
    l.bal.values().map(|r| r.get(res).copied().unwrap_or(0)).sum()
}

// the one primitive: move `amt` of `res` between two accounts. Conserves the grand total.
// Returns None for INFEASIBLE (no overdraft, no negative amount).
pub fn transfer(l: &Ledger, res: &str, from: &str, to: &str, amt: i64) -> Option<Ledger> {
    if amt < 0 || balance(l, from, res) < amt {
        return None;
    }
    let mut m = l.clone();
    let fb = m.bal.entry(from.to_string()).or_default();
    let cur_from = fb.get(res).copied().unwrap_or(0);
    fb.insert(res.to_string(), cur_from - amt);
    let tb = m.bal.entry(to.to_string()).or_default();
    let cur_to = tb.get(res).copied().unwrap_or(0);
    tb.insert(res.to_string(), cur_to + amt);
    Some(m)
}

pub fn spend(l: &Ledger, acct: &str, res: &str, amt: i64) -> Option<Ledger> {
    transfer(l, res, acct, SINK, amt)
}
pub fn refill(l: &Ledger, acct: &str, res: &str, amt: i64) -> Option<Ledger> {
    transfer(l, res, TREASURY, acct, amt)
}

pub fn affords(l: &Ledger, acct: &str, cost: &[(&str, i64)]) -> bool {
    cost.iter().all(|(res, amt)| balance(l, acct, res) >= *amt)
}
pub fn feasible(l: &Ledger, acct: &str, cost: &[(&str, i64)]) -> bool {
    affords(l, acct, cost)
}

pub struct UseResult {
    pub ok: bool,
    pub l: Ledger,
}

// reusable (`!`) vs depletable: `use` depletes a depletable resource, but never a reusable one
pub fn use_res(l: &Ledger, acct: &str, res: &str) -> UseResult {
    if balance(l, acct, res) < 1 {
        return UseResult { ok: false, l: l.clone() };
    }
    if l.kind.get(res).map(|s| s.as_str()) == Some("reusable") {
        return UseResult { ok: true, l: l.clone() };
    }
    UseResult {
        ok: true,
        l: spend(l, acct, res, 1).unwrap_or_else(|| l.clone()),
    }
}

// continual learning: capacity is conserved; knowledge is reusable.
pub fn allocate(l: &Ledger, task: &str, amt: i64) -> Option<Ledger> {
    transfer(l, "capacity", FREE, &format!("task:{}", task), amt)
}

pub fn consolidate(l: &Ledger, task: &str, mind: &str) -> Ledger {
    let mut m = l.clone();
    m.kind.insert(format!("know:{}", task), "reusable".to_string());
    let mb = m.bal.entry(mind.to_string()).or_default();
    mb.insert(format!("know:{}", task), 1);
    m
}

pub fn forget(l: &Ledger, task: &str, mind: &str) -> Ledger {
    let amt = balance(l, &format!("task:{}", task), "capacity");
    let mut m = transfer(l, "capacity", &format!("task:{}", task), FREE, amt)
        .unwrap_or_else(|| l.clone());
    if m.bal.contains_key(mind) {
        m.bal.get_mut(mind).unwrap().insert(format!("know:{}", task), 0);
    }
    m
}

// PRICING THE REPAIRS (Type II rationality)
pub fn worthwhile(value: i64, cost: i64) -> bool {
    value >= cost
}

pub struct RepairResult {
    pub decision: &'static str, // "cannot-afford" | "skip" | "invoke"
    pub l: Ledger,
}

pub fn repair(l: &Ledger, acct: &str, resource: &str, value: i64, cost: i64) -> RepairResult {
    if !affords(l, acct, &[(resource, cost)]) {
        return RepairResult { decision: "cannot-afford", l: l.clone() };
    }
    if !worthwhile(value, cost) {
        return RepairResult { decision: "skip", l: l.clone() };
    }
    RepairResult {
        decision: "invoke",
        l: spend(l, acct, resource, cost).unwrap_or_else(|| l.clone()),
    }
}
